Authentication of Oracle API endpoints

elijaholmos opened the pull request:
## summary
Implements API authentication for all Oracle endpoints. All non-GET methods require authentication. Furthermore, any attempts to modify existing data can only be performed by the owner of that data (the `oracleId` connected to the apikey must match the `oracleId` of the resource attempting to be modified)
A master API key is connected to `oracleId` 0 (a SYSTEM account) and has full access to all operations. It is the only allowed form of authorization for POSTing the `/oracle` endpoint.
API keys should be provided in the Authorization header of the HTTP request and follow the HTTP Bearer Authentication standard.

## todo
- Once the current implementation for Oracle endpoints is validated, I will apply a similar implementation for the other endpoints
- Expand the `AuthError` class and integrate it with the `ErrorResponse` model

elijaholmos committed:
remove commented import

elijaholmos committed:
setup auth backend w/o implementing

elijaholmos committed:
check before redoing implementation

elijaholmos committed:
first refactor - boilerplate, ready for customization

this actually somewhat works?? wow

elijaholmos committed:
checkpoint: apiKey auth works, next: detailed 401/403

elijaholmos committed:
nearly complete

elijaholmos committed:
npm audit

elijaholmos committed:
cleanup middleware, detect verifyFn bug

kharaone commented:
@elijaholmos I will need some help in order to reproduce your error and setup the whole api with auth

elijaholmos committed:
correct oracle id

kharaone committed:
Merge branch 'dev' into endpoint-authentication

kharaone commented:
@elijaholmos ok, it seems the error is tied to the connector : 
it's using `Object.prototype.hasOwnProperty.call(where, 'key')` but it's incorrectly setting the property name, instead of using the value contained in the variable, it's using the string name of the variable.
if I switch the `Object.prototype.hasOwnProperty.call(where, 'key')` to `Object.prototype.hasOwnProperty.call(where, key)` on line 460 of `src\connectors\BetterFirestoreConnector.ts` I'm getting through correctly.

feel free to review the connector code for these specific cases

kharaone reviewed (changes requested):
see previous comment

elijaholmos committed:
Update BetterFirestoreConnector.ts

elijaholmos commented:
@kharaone I verified that your proposed solution fixed the bug - thank you for helping me solve it
PR description has been updated accordingly and is ready for the next stage of review

kharaone committed:
fix: extending good pattern to other part of the code

- Object.prototype.hasOwnProperty.call requires the name of the key, it wasn't provided with the correct one

kharaone commented on src/sequence.ts:
```diff
@@ -1,3 +1,75 @@
-export class MySequence extends MiddlewareSequence {}
+const SequenceActions = RestBindings.SequenceActions;
+
+export class MySequence implements SequenceHandler {
```
I would have mentioned the fact that the middleware route seemed to be the correct way to go, but seeing you removed the ` extends MiddlewareSequence` makes me wonder why you chose to go back to the sequenceHandler ?

elijaholmos replied:
It was largely motivated by lb4's example found at https://github.com/loopbackio/loopback-next/blob/e2ff6b22367e919926d0f41f6d939d988c654c00/examples/todo-jwt/src/sequence.ts

kharaone commented on src/sequence.ts:
```diff
@@ -1,3 +1,75 @@
+      if (request.method === 'POST' && request.body?.id != oracle.id)
+        throw new AuthError(403);
+      //only owners can PUT requests
+      if (route.pathParams.id != oracle.id) throw new AuthError(403);
```
This part is specific to /oracle endpoint and will need to be constrained accordingly. Moreover, it feels more like authorization than authentication, you will need to find a better place to define these rules

elijaholmos replied:
refactored in 8e71216344520cba30854a9c3ded48664ded42bd

kharaone commented on src/sequence.ts:
```diff
@@ -1,3 +1,75 @@
+      const oracle: GoodOracle = await this.authenticateRequest(request);
+      console.log(oracle);
+      // oracle is undefined only when @authenticate.skip() is used
+      if (oracle === undefined) return next();
```
needs to be awaited

kharaone commented on src/sequence.ts:
```diff
@@ -1,3 +1,75 @@
+      if (oracle === undefined) return next();
+
+      // extra auth checks
+      if (oracle.id == 0) return next(); // SYSTEM account
```
needs to be awaited

kharaone commented on src/sequence.ts:
```diff
@@ -1,3 +1,75 @@
+      if (oracle.id == 0) return next(); // SYSTEM account
+      //only owners can POST requests
+      if (request.method === 'POST' && request.body?.id != oracle.id)
+        throw new AuthError(403);
```
if if's not a one liner, put the {} around the if scope

kharaone commented on src/providers/verifyFn.provider.ts:
```diff
@@ -0,0 +1,48 @@
+          where: {
+            apikey: token,
+          },
+        };
```
Shouldn't we have a check on expired too ?

kharaone commented on src/providers/verifyFn.provider.ts:
```diff
@@ -0,0 +1,48 @@
+    const self = this;
+
+    return async function (token: string, cb: Function) {
+      console.log('in value! token:', token);
```
to be removed

kharaone commented on src/providers/verifyFn.provider.ts:
```diff
@@ -0,0 +1,48 @@
+            apikey: token,
+          },
+        };
+        console.log(f);
```
to be removed

kharaone commented on src/providers/verifyFn.provider.ts:
```diff
@@ -0,0 +1,48 @@
+        };
+        console.log(f);
+        const oracleApiKey = await self.authRepository.findOne(f);
+        console.log(oracleApiKey);
```
to be removed

kharaone commented on src/providers/verifyFn.provider.ts:
```diff
@@ -0,0 +1,48 @@
+          oracleApiKey?.apikey !== token ||
+          !!oracleApiKey?.expired
+        )
+          throw new AuthError(401);
```
if not a oneliner, please encapsulate with `{}`

kharaone commented on src/oracleprofile.factory.ts:
```diff
@@ -0,0 +1,9 @@
+export const OracleProfileFactory: UserProfileFactory<GoodOracle> = function (
+  oracle: GoodOracle,
+): GoodOracle {
+  return {...oracle, [securityId]: oracle?.id?.toString() || ''} as GoodOracle;
```
```suggestion
  return {...oracle, [securityId]: oracle?.id?.toString() ?? ''} as GoodOracle;
```

kharaone commented on src/oracleprofile.factory.ts:
```diff
@@ -0,0 +1,9 @@
+import {securityId} from '@loopback/security';
+import {GoodOracle} from './models';
+
+export const OracleProfileFactory: UserProfileFactory<GoodOracle> = function (
```
I'd move that file it in models

kharaone commented on src/models/good-oracle.model.ts:
```diff
@@ -59,6 +60,11 @@ export class GoodOracle extends Entity {
+  @property({
+    hidden: true,
+  })
+  [securityId]: string = this.id?.toString() || '';
```
```suggestion
  [securityId]: string = this.id?.toString() ?? '';
```

kharaone commented on src/providers/auth-error.ts:
```diff
@@ -0,0 +1,9 @@
+export default class AuthError extends Error {
```
agreed, this should be translated into an errorresponse either through an interceptor or a middleware

kharaone commented on src/providers/passport-bearer-auth.ts:
```diff
@@ -0,0 +1,40 @@
+import {Strategy, VerifyFunction} from 'passport-http-bearer';
+
+export const AUTH_STRATEGY_NAME = 'bearer';
+export class PassportBearerAuthProvider<GoodOracle>
```
should probably contain `provider `in the filename to stay consistent in the naming convention

kharaone reviewed (changes requested):
Here are a few suggestions on how to proceed further

kharaone committed:
Merge branch 'dev' into endpoint-authentication

elijaholmos committed:
code cleanup per PR suggestions

kharaone reviewed (approved):
I'm still concerned of the part in sequence.ts which seemed specific to /oracle endpoint. You should at least guard the section with a conditional on the url 

if url === /oracle then apply rules.

elijaholmos commented:
@kharaone sorry, that was only the first part 😅 I'll re-request review once I'm finished with all the suggested changes

elijaholmos committed:
check for expired API key in query

elijaholmos committed:
consolidate route authorization by endpoint

elijaholmos commented:
@kharaone two final questions:
1. for the sake of PR scope, could a separate story be created for the handling & displaying of API errors to the requester? including converting `AuthError`s into `ErrorResponse`s?
2. would you like me to implement authentication on the remaining endpoints before merging this? `/activity`, `/entry`, etc

kharaone commented:
@elijaholmos 
1. sure, we can implement that in another story indeed
2. you should avoid /entry for now as @gitluky is getting that part overhauled. Feel free to either create separate stories for each endpoints or implement these in that PR

elijaholmos commented:
@kharaone thank you, I will go ahead and move everything to separate stories so this can be merged once re-approved

elijaholmos committed:
refactor to use @loopback/authorization

kharaone committed:
lint: refactored different part of the codebase to fix lint issues

- switch to operation decorator instead of using @operation (eg, @get)
- removed ts-expect-error for better typings
- return in invalidateLastApiKey was smelly
- sequence needed some awaiting next

kharaone reviewed (approved):
- I didn't update the tests with the authentication dimension (as there's apparently no way to insert an ID = 0 in the datasource, so we'll need to change the admin flag to be an actual flag)
- any person calling /auth with oracleId = X can "override" its apikey ... feels a bit problematic. We will probably need to do secret ids or even jwt token flow instead.

elijaholmos commented:
I have a few ideas about the second point, for discussion at a future sync
